// react-md/es/Cards/CardMedia.js
var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';

import deprecated from 'react-prop-types/lib/deprecated';
import componentDeprecated from '../utils/PropTypes/componentDeprecated';
import Media from '../Media/Media';
import MediaOverlay from '../Media/MediaOverlay';

/**
 * The `CardMedia` component is used to display images or some sort
 * media.
 */

var CardMedia = function (_PureComponent) {
  _inherits(CardMedia, _PureComponent);

  function CardMedia() {
    _classCallCheck(this, CardMedia);

    return _possibleConstructorReturn(this, _PureComponent.apply(this, arguments));
  }

  CardMedia.prototype.render = function render() {
    var _props = this.props,
        className = _props.className,
        children = _props.children,
        props = _objectWithoutProperties(_props, ['className', 'children']);

    delete props.overlay;

    var overlay = this.props.overlay;

    if (overlay) {
      overlay = React.createElement(
        MediaOverlay,
        null,
        overlay
      );
    }

    return React.createElement(
      Media,
      _extends({ className: cn('md-card-media', className) }, props),
      children,
      overlay
    );
  };

  return CardMedia;
}(PureComponent);

CardMedia.aspect = {
  equal: '1-1',
  wide: '16-9'
};
CardMedia.propTypes = {
  /**
   * An optional className to apply to the card media component.
   */
  className: PropTypes.string,

  /**
   * An optional overlay component to be rendered over the media. This *should*
   * be A `CardTitle`, `CardActions` or both.
   */
  overlay: deprecated(PropTypes.node, 'Use the `MediaOverlay` component as a child instead'),

  /**
   * Any media to display.
   */
  children: PropTypes.node,

  /**
   * Boolean if the aspect ratio should be forced.
   */
  forceAspect: PropTypes.bool,

  /**
   * The aspect ratio to use.
   */
  aspectRatio: PropTypes.oneOf([CardMedia.aspect.equal, CardMedia.aspect.wide]).isRequired,

  /**
   * Boolean if this component should be expandable when there is a `CardExpander`
   * above it in the `Card`.
   */
  expandable: PropTypes.bool,

  /**
   * The component to render the card media as.
   */
  component: PropTypes.oneOfType([PropTypes.func, PropTypes.string, PropTypes.object]).isRequired,

  deprecated: componentDeprecated('There were no unique styles for media in cards so it is simpler to just use the ' + '`Media` component.')
};
CardMedia.defaultProps = {
  forceAspect: true,
  aspectRatio: CardMedia.aspect.wide,
  component: 'section'
};
export default CardMedia;
