// react-md/es/Pickers/PickerFooter.js
function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';

import Button from '../Buttons/Button';

/**
 * The `PickerFooter` component is a dialog footer for the `Picker` components.
 * It renders an ok and cancel button.
 */

var PickerFooter = function (_PureComponent) {
  _inherits(PickerFooter, _PureComponent);

  function PickerFooter() {
    _classCallCheck(this, PickerFooter);

    return _possibleConstructorReturn(this, _PureComponent.apply(this, arguments));
  }

  PickerFooter.prototype.render = function render() {
    var _props = this.props,
        className = _props.className,
        okLabel = _props.okLabel,
        okPrimary = _props.okPrimary,
        onOkClick = _props.onOkClick,
        cancelLabel = _props.cancelLabel,
        cancelPrimary = _props.cancelPrimary,
        onCancelClick = _props.onCancelClick;


    return React.createElement(
      'footer',
      { className: cn('md-dialog-footer', className) },
      React.createElement(Button, {
        flat: true,
        primary: cancelPrimary,
        secondary: !cancelPrimary,
        label: cancelLabel,
        onClick: onCancelClick
      }),
      React.createElement(Button, {
        flat: true,
        primary: okPrimary,
        secondary: !okPrimary,
        label: okLabel,
        onClick: onOkClick
      })
    );
  };

  return PickerFooter;
}(PureComponent);

PickerFooter.propTypes = {
  className: PropTypes.string,
  okLabel: PropTypes.node.isRequired,
  okPrimary: PropTypes.bool.isRequired,
  onOkClick: PropTypes.func.isRequired,
  cancelLabel: PropTypes.node.isRequired,
  cancelPrimary: PropTypes.bool.isRequired,
  onCancelClick: PropTypes.func.isRequired
};
export default PickerFooter;
